### PR TITLE
ARROW-8802: [C++][Dataset] Preserve dataset schema's metadata on column projection

### PR DESCRIPTION
**Pull request by fsaintjacques**
Scanner does not preserve the original schema metadata when columns are projected.

### CODE REVIEW

**Comment by fsaintjacques** [comment]
@jorisvandenbossche I'm puzzled by the error. Could it be an upstream pandas change?

**Comment by jorisvandenbossche** [comment]
So this was actually a "time bomb" in the extension type testing code: we were using a test extension type with the same name than an extension type that in the meantime has been implemented in pandas. So depending on whether pandas already registered this type with pyarrow or not, would trigger this error "A type extension with name pandas.period already defined". 

And using the pandas parquet functionality, triggers pandas to register this extension type. So until now, when running the tests in normal order, no test has been using pandas' parquet functions, before testing `test_extension_type.py`. But with this PR, a test using pandas' `to_parquet` was dded to `test_dataset.py`, which is run before `test_extension_type.py` ..

If I explicitly tell pytest to run eg ``test_parquetpy`` first and then `test_extension_type.py`, I get the exact same failure on master.

TLDR: pushed a fix to rename our test extension type class ;)

File: python/pyarrow/tests/test_dataset.py [hunk 1]
```diff
@@ -1566,3 +1566,21 @@ def test_parquet_dataset_factory_partitioned(tempdir):
...
+    df = pd.DataFrame({'a': [1, 2, 3]})
+    path = tempdir / "test.parquet"
+    df.to_parquet(path)
+    dataset = ds.dataset(path)
+
+    schema = dataset.to_table().schema
+    projected_schema = dataset.to_table(columns=["a"]).schema
+
+    # ensure the pandas metadata is included in the schema
+    assert b"pandas" in schema.metadata
```
Comment (jorisvandenbossche) [inline python/pyarrow/tests/test_dataset.py:1584]: added an additional assert to ensure the "pandas" key is actually present, because if for some reason we accidentally remove it in both cases, this test won't detect that as both schema's are still identical (both missing the metadata). 
(although I assume we have other tests that would start failing then)

**Review by jorisvandenbossche — APPROVED** [review]
